declaration extraction

jermy-c opened the pull request:
**Completed** 
- basic global declaration extraction
- basic struct declaration extraction
- basic function declaration extraction
- basic unit testing
- remove comment function
- extract package function

**Issues still need to be resolved**

- [x] alias and enum need to be added
- [x] possible func() or func ()
- [x] better unit and integration testing
- [x] possibility of comments in quotes which should not be comment code, example: method("//", argument)
- [x] comments/docs

jermy-c committed:
start declaration extraction

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/skycoin/cx?pullRequest=772) <br/>All committers have signed the CLA.

kenje4090 commented on cmd/declaration_extraction/declaration_extraction_test.go:
```diff
@@ -0,0 +1,125 @@
+		Declaration{
+			PackageID:   "hello",
+			FileID:      "test.cx",
+			StartOffset: 0,
+			Length:      3,
```
Can you pls double-check, I don't think these values are the right ones. for `var apple string` the length should be 16. So when we have an offset, we can get the whole declaration by using `sourceCodeBytes[offset:offset+length]`

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,297 @@
+	for scanner.Scan() {
+		line := scanner.Bytes()
+
+		commentLoc := reComment.FindIndex(line)
+
+		multiCommentOpenLoc := reMultiCommentOpen.FindIndex(line)
+		multiCommentCloseLoc := reMultiCommentClose.FindIndex(line)
+
+		//Check for commented code
+		if commentedCode && multiCommentCloseLoc != nil {
+			commentedCode = false
+		}
+
+		if commentedCode {
+			continue
+		}
+
+		if multiCommentOpenLoc != nil && !commentedCode && multiCommentCloseLoc != nil {
+			commentedCode = true
+		}
```
All functions right now are manually checking comments. Let's do this instead, first get the source code, then remove all the comments, and then the source code that has no comments will be passed to these extractor functions.

jermy-c committed:
add rmComment and extractPkg

jermy-c committed:
add extractEnum

-bugs to resolve offset

jermy-c committed:
start restructure Tests

jermy-c committed:
fix enum and offest

- finish extract enum function
- restructure unit tests
- fix offset
- start redeclaration check

kenje4090 commented on cmd/declaration_extraction/declaration_extraction_test.go:
```diff
@@ -0,0 +1,231 @@
+	"testing"
+)
+
+func TestDeclarationExtraction(t *testing.T) {
```
Please separate the tests for each functions. It should be TestDeclarationExtraction_ExtractGlobals(),TestDeclarationExtraction_ExtractStructs(), TestDeclarationExtraction_ExtractEnums() etc

kenje4090 commented on cmd/declaration_extraction/declaration_extraction_test.go:
```diff
@@ -0,0 +1,231 @@
+const (
+	First Number = iota
+	Second
+)`,
```
You have test.cx that contains this. You can use that one instead of manually putting it here.

kenje4090 commented on cmd/declaration_extraction/declaration_extraction_test.go:
```diff
@@ -0,0 +1,231 @@
+package declaration_extraction
```
Pls make this `package declaration_extraction_test`

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,313 @@
+	return src
+}
+
+func extractPkg(source []byte) string {
```
Pls change this to `ExtractPackages`, so it will be public and can be imported

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,313 @@
+	return srcStr
+}
+
+func extractGlbl(source *os.File) []Declaration {
```
Pls change this to `ExtractGlobals`, do this with the other extract functions too

jermy-c committed:
Merge branch 'skycoin:develop' into declaration-extraction

jermy-c committed:
reDeclar func and tests

jermy-c committed:
Merge branch 'declaration-extraction' of https://github.com/jermy-c/cx into declaration-extraction

kenje4090 commented on cmd/declaration_extraction/go.mod:
```diff
@@ -0,0 +1,3 @@
+module test.com/declaration_extraction
+
+go 1.16
```
Pls remove this file

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,306 @@
+	Name        string
+}
+
+func rmComment(source []byte) []byte {
```
Just name this to `removeComment()`

kenje4090 replied:
Also, pls add unit test for this

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,306 @@
+	var pkgID string
+	src, err := io.ReadAll(source)
+
+	CmtRmd := rmComment(src)
```
There should be no removeComment() calls inside the extract functions, what should happen is we pass the source code to removeComment() then the source code without comments will be passed to extract functions

kenje4090 commented on cmd/declaration_extraction/declaration_extraction_test.go:
```diff
@@ -0,0 +1,395 @@
+	"os"
+	"testing"
+
+	"test.com/declaration_extraction"
```
Why is this "test.com/declaration_extraction"

kenje4090 reviewed (commented):
You can also put the .cx test files under a folder like "test_files/redeclaration_enum.cx"

jermy-c committed:
pass rmComment to extract functions

jermy-c committed:
fix func prototype with error

jermy-c committed:
rename to reComment

jermy-c committed:
fix ExtractPackages test to use removedcomment src

jermy-c committed:
replace comments with whitespace

jermy-c committed:
fix offset struct

jermy-c committed:
global offset fix

jermy-c committed:
fix offset for rest of funcs

jermy-c committed:
fix regex for func

jermy-c committed:
fix func being deleted

jermy-c committed:
fix tests

jermy-c committed:
fix getdeclaration test

kenje4090 commented on cmd/declaration_extraction/declaration_extraction_test.go:
```diff
@@ -0,0 +1,546 @@
+			if err != nil {
+				t.Fatal(err)
+			}
+			commentReplaced := declaration_extraction.ReplaceCommentsWithWhitespaces(srcBytes)
```
Can you also check if the length of srcBytes is equal to the length of commentReplaced? And the line number?

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,298 @@
+	"unicode"
+)
+
+type GlobalDeclaration struct {
+	PackageID   string
+	FileID      string
+	StartOffset int
+	Length      int
+	Name        string
+}
+
+type EnumDeclaration struct {
+	PackageID   string
+	FileID      string
+	StartOffset int
+	Length      int
+	Name        string
+	Type        string
+	Value       int
+}
+
+type StructDeclaration struct {
+	PackageID   string
+	FileID      string
+	StartOffset int
+	Length      int
+	Name        string
+}
+
+type FuncDeclaration struct {
+	PackageID   string
+	FileID      string
+	StartOffset int
+	Length      int
+	Name        string
+}
```
Please add LineNumber int field to all

jermy-c committed:
add line number

jermy-c committed:
start fixing offset for extractEnum

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,323 @@
+	StartOffset int
+	Length      int
+	LineNumber  int
+	Name        string
```
Also please change `Name` fields to `GlobalVariableName`, and the other structs with name field too. Also add comment beside it `// name of variable being declared`

jermy-c committed:
fix enum offset

jermy-c committed:
fix struct offset

jermy-c committed:
fix offset for the rest

jermy-c committed:
change Name to [declararion]VariableName

jermy-c committed:
fix func dec offset

jermy-c committed:
start extract all func

jermy-c committed:
add concurrency

kenje4090 commented on cmd/declaration_extraction/declaration_extraction.go:
```diff
@@ -0,0 +1,431 @@
+	Funcs := <-FuncsCh
+
+	if err := ReDeclarationCheck(Globals, Enums, Structs, Funcs); err != nil {
+		log.Fatal(err)
```
also, pls return this error, don't just log.Fatal it

jermy-c committed:
start nested goroutines

jermy-c committed:
fix wg sleeping

jermy-c committed:
new concurrent process

jermy-c committed:
attempt to fix waitgroup

jermy-c committed:
pass all params to funcs

jermy-c committed:
Merge branch 'skycoin:develop' into declaration-extraction

jermy-c committed:
add docs

jermy-c committed:
fix sleeping

jermy-c committed:
complete extractall func

jermy-c committed:
start extractAlltest

jermy-c committed:
start multiple file test

jermy-c committed:
fix buffered channel recieving

jermy-c committed:
fixed func reacieving

jermy-c committed:
continue writing tests

jermy-c committed:
add scanlines func

jermy-c committed:
fix offest for mac/linux

jermy-c committed:
complete test scenario

jermy-c committed:
change tests

jermy-c committed:
add scenarios

jermy-c committed:
fix tests

jermy-c committed:
add mulitple packages detection

jermy-c committed:
Revert "add mulitple packages detection"

This reverts commit 15a9db32c286a91bcc03e0856e36580d56979a82.

jermy-c committed:
add test_2 scenario

jermy-c committed:
read only first line for extract packages

jermy-c committed:
add comments

jermy-c committed:
add final comments
